### ISSUE TITLE
[BUG] Mask expansion / inflated masks with Cellpose-SAM on single-channel microscopy images

### ISSUE THREAD

**Issue by walkeraw2002** [MouseLand/cellpose#1442]
### **The Bug**
When segmenting single-channel microscopy images with Cellpose-SAM, I’m seeing mask expansion / inflated masks. There is expansion for both membrane markers and nuclear markers.

When I triplicate the channel to RGB, I see improved results. When I use an RGB screenshot taken from an image viewer, I also see improved results. 

I have tried cellpose default normalization along with custom normalization methods. I have tried changing image bit depth as well, but the only thing that changes the behavior is triplicating the single channel.

In the docs it states that Cellpose-SAM was trained on two channel images (one membrane and one nuclei), with the other channel set to zero, but I'm not sure what the best practices are for segmenting membrane or nuclei on it's own as a single channel. 

Below is an example where we see mask expansion when segmenting a CD8 membrane image. To generate this result, CD8 was placed in the first channel and the other two were zeroed out:

<img width="1228" height="1436" alt="Image" src="https://github.com/user-attachments/assets/ff48f8f3-1933-42d1-9e33-6f9e293f9d87" />
<br>
<br>

This is the result from triplicating CD8 to occupy all three channels before segmenting the image, where we don't see mask expansion:

<img width="1229" height="1386" alt="Image" src="https://github.com/user-attachments/assets/80953cdf-3f89-419e-bb99-2f7d03f802b5" />
<br>
<br>

We also see mask expansion when segmenting DAPI single channel images (although the issue is less prevalent):

<img width="1244" height="931" alt="Image" src="https://github.com/user-attachments/assets/d0ee3c89-00e3-4424-b5e8-b5972928f8ca" />
<br>
<br>

### **Question**
Are there any tips or best practices out there for segmenting single channel images and avoiding this mask expansion? Is the only way to triplicate the single channel across all three channels or are there alternative ways of solving this issue? I'm not sure if the model being trained on images with cytoplasm and nuclear channels in any order, and the other channel set to zero, is causing this to happen.

Any advice would be much appreciated. Thank you for your time.

**Comment by mrariden** [comment]
Increasing the `cellprob` threshold will make your cells smaller. You will likely have to tune this parameter for your data

**Comment by cyh5757** [comment]
> Increasing the `cellprob` threshold will make your cells smaller. You will likely have to tune this parameter for your data

Hi, @mrariden.  thank you for your great work.

I was wondering if there are any other ways to adjust cell contours besides changing the cellprob threshold.

Since inference is slow on large images, I am currently using a workflow where I resize the images to smaller sizes before running the model. However, when I resize the images and adjust the cellprob threshold, the contours, or masks, are extracted in a way that fits the cells better, but small cells tend to disappear as a result.

Are there any recommended approaches in terms of training strategy or post-processing to address this issue?

The image sizes in my dataset are approximately 2200 × 2200, 1900 × 1400, and 1900 × 1900. I first convert each image into a square format, adding padding when the image is not square, and then resize it to 512 × 512. The model input and output are both 192 × 192. During inference, the output is resized back to the original input image size.

I would appreciate any suggestions or advice.

@walkeraw2002 Hi, have you tried any other approaches besides adjusting the cellprob threshold?

**Comment by mrariden** [comment]
> I was wondering if there are any other ways to adjust cell contours besides changing the cellprob threshold.

Not internally within cellpose. Putative pixels belonging to a cell are predicted by the network and thresholded by the cellprob threshold. Decreasing the threshold will omit fewer pixels which tend to be at the edge of each cell. 

> Are there any recommended approaches in terms of training strategy or post-processing to address this issue?

You could dilate cell masks in post-processing. If you wanted to fine-tune a cp model, you could use these dilated masks as training data.

**Comment by cyh5757** [comment]
@mrariden Thank you for the clarification and suggestion.

My main goal was to improve this behavior through the training objective rather than relying on additional pre-processing or post-processing. A more data-centric approach, such as generating dilated masks and using them as training targets, would likely require more time in my current workflow.

Instead, I modified the Cellpose-SAM training loss directly. In particular, I replaced the original cellprob loss with focal loss and trained the model using that change. The improvement was not large, but it was meaningful in my experiments, especially in terms of preserving smaller cells while improving the mask fit.

I will keep experimenting with this direction. Thank you again for the helpful suggestion.

@walkeraw2002 I improved by using focal loss in deciding of cellprob. Good luck to you.
